server/src/main/java/org/opensearch/snapshots/SnapshotsService.java

```diff
@@ -407,6 +424,177 @@ public TimeValue timeout() {
+     * operation.
+     *
+     * @param request  snapshot request
+     * @param listener snapshot creation listener
+     */
+    public void createSnapshotV2(final CreateSnapshotRequest request, final ActionListener<SnapshotInfo> listener) {
+        long pinnedTimestamp = System.currentTimeMillis();
+        final String repositoryName = request.repository();
+        final String snapshotName = indexNameExpressionResolver.resolveDateMathExpression(request.snapshot());
+        validate(repositoryName, snapshotName);
+
+        final SnapshotId snapshotId = new SnapshotId(snapshotName, UUIDs.randomBase64UUID()); // new UUID for the snapshot
+        Repository repository = repositoriesService.repository(request.repository());
+
+        if (repository.isReadOnly()) {
+            listener.onFailure(
+                new RepositoryException(repository.getMetadata().name(), "cannot create snapshot-v2 in a readonly repository")
+            );
+            return;
+        }
+
+        final Snapshot snapshot = new Snapshot(repositoryName, snapshotId);
+        ClusterState currentState = clusterService.state();
+        final Map<String, Object> userMeta = repository.adaptUserMetadata(request.userMetadata());
+        try {
+            final StepListener<RepositoryData> repositoryDataListener = new StepListener<>();
+            repositoriesService.getRepositoryData(repositoryName, repositoryDataListener);
+
+            repositoryDataListener.whenComplete(repositoryData -> {
+                ensureSnapshotNameAvailableInRepo(repositoryData, snapshotName, repository);
+                final SnapshotsInProgress snapshots = currentState.custom(SnapshotsInProgress.TYPE, SnapshotsInProgress.EMPTY);
+                final List<SnapshotsInProgress.Entry> runningSnapshots = snapshots.entries();
+                ensureSnapshotNameNotRunning(runningSnapshots, repositoryName, snapshotName);
+                validate(repositoryName, snapshotName, currentState);
+                final SnapshotDeletionsInProgress deletionsInProgress = currentState.custom(
+                    SnapshotDeletionsInProgress.TYPE,
+                    SnapshotDeletionsInProgress.EMPTY
+                );
+                final RepositoryCleanupInProgress repositoryCleanupInProgress = currentState.custom(
+                    RepositoryCleanupInProgress.TYPE,
+                    RepositoryCleanupInProgress.EMPTY
+                );
+                if (repositoryCleanupInProgress.hasCleanupInProgress()) {
+                    throw new ConcurrentSnapshotExecutionException(
+                        repositoryName,
+                        snapshotName,
+                        "cannot snapshot-v2 while a repository cleanup is in-progress in [" + repositoryCleanupInProgress + "]"
+                    );
+                }
+                ensureNoCleanupInProgress(currentState, repositoryName, snapshotName);
```

Comment:
We won't be supporting multiple concurrent create operations in snapshot v2. At present too, concurrent snapshot calls are queued. Since the snapshot v2 call is sync, we don't see need for queuing. The client can retry the call if it fails.

Reply:
Let's assume there is a 1. create or 2. delete snapshot event ongoing. What would be the behaviour if there is  client call to a) create snapshot or b) delete snapshot.